File: vigia-web/src/app/core/services/realtime/device-groups-realtime.service.ts
import { HttpBackend, HttpClient } from '@angular/common/http';
import { inject, Injectable, OnDestroy } from '@angular/core';
import {
  GroupMembershipChangedDto,
  parseGroupMembershipChanged,
} from '@core/entities';
import { AuthSessionService } from '@core/services/auth/auth-session.service';
import { environment } from '@environments/environment';
import {
  HubConnection,
  HubConnectionBuilder,
  HubConnectionState,
} from '@microsoft/signalr';
import { AuthTokensDto } from '@core/entities/DTOs/auth.dto';
import { Subject, firstValueFrom } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class DeviceGroupsRealtimeService implements OnDestroy {
  private readonly session = inject(AuthSessionService);
  private readonly refreshHttp = new HttpClient(inject(HttpBackend));

  private connection: HubConnection | null = null;
  private readonly membershipChangedSubject = new Subject<GroupMembershipChangedDto>();

  readonly membershipChanged$ = this.membershipChangedSubject.asObservable();

  ngOnDestroy(): void {
    void this.disconnect();
    this.membershipChangedSubject.complete();
  }

  get isConnected(): boolean {
    return this.connection?.state === HubConnectionState.Connected;
  }

  async connect(): Promise<void> {
    const current = this.connection;
    if (
      current &&
      (current.state === HubConnectionState.Connected ||
        current.state === HubConnectionState.Connecting ||
        current.state === HubConnectionState.Reconnecting)
    ) {
      return;
    }

    await this.disconnect();

    const token = await this.ensureAccessToken();
    if (!token) {
      return;
    }

    const hubUrl = `${environment.apiUrl.replace(/\/$/, '')}/hubs/device-groups`;
    const connection = new HubConnectionBuilder()
      .withUrl(hubUrl, {
        accessTokenFactory: () => this.ensureAccessToken(),
      })
      .withAutomaticReconnect([0, 2000, 5000, 10000, 30000])
      .build();

    connection.on('GroupMembershipChanged', (...args: unknown[]) => {
      const event = parseGroupMembershipChanged(args[0]);
      if (event) {
        this.membershipChangedSubject.next(event);
      }
    });

    this.connection = connection;
    await connection.start();
  }

  async disconnect(): Promise<void> {
    const connection = this.connection;
    this.connection = null;
    if (!connection) {
      return;
    }

    try {
      connection.off('GroupMembershipChanged');
      await connection.stop();
    } catch {
      // Ignore teardown errors.
    }
  }

  private async ensureAccessToken(): Promise<string> {
    const current = this.session.getAccessToken();
    if (current) {
      return current;
    }

    const refreshToken = this.session.getRefreshToken();
    if (!refreshToken) {
      return '';
    }

    try {
      const base = environment.apiUrl.replace(/\/$/, '');
      const tokens = await firstValueFrom(
        this.refreshHttp.post<AuthTokensDto>(`${base}/auth/refresh`, { refreshToken }),
      );
      if (tokens) {
        this.session.setSession(tokens);
        return tokens.accessToken;
      }
    } catch {
      // Fall through.
    }

    return current ?? '';
  }
}

File: vigia_ui/lib/data/services/device_groups_realtime_service.dart
import 'dart:async';

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:jwt_decoder/jwt_decoder.dart';
import 'package:signalr_netcore/signalr_client.dart';
import 'package:vigia_ui/data/services/token_storage_service.dart';
import 'package:vigia_ui/domain/DTOs/group_membership_changed.dart';
import 'package:vigia_ui/domain/environments.dart';

typedef GroupMembershipHandler = void Function(GroupMembershipChanged event);

/// Thin wrapper around the ASP.NET Core SignalR device-groups hub.
class DeviceGroupsRealtimeService {
  DeviceGroupsRealtimeService({
    required this._tokenStorage,
    required this._refreshDio,
  });

  final TokenStorageService _tokenStorage;
  final Dio _refreshDio;
  static const _hubPath = '/hubs/device-groups';
  static const _membershipChangedEvent = 'GroupMembershipChanged';
  HubConnection? _connection;
  final _handlers = <GroupMembershipHandler>{};

  bool get isConnected => _connection?.state == HubConnectionState.Connected;

  void addHandler(GroupMembershipHandler handler) => _handlers.add(handler);

  void removeHandler(GroupMembershipHandler handler) =>
      _handlers.remove(handler);

  Future<void> connect() async {
    if (_connection != null &&
        (_connection!.state == HubConnectionState.Connected ||
            _connection!.state == HubConnectionState.Connecting ||
            _connection!.state == HubConnectionState.Reconnecting)) {
      return;
    }

    await disconnect();

    final token = await _ensureAccessToken();
    if (token.isEmpty) {
      debugPrint('[SignalR] skip connect: no access token');
      return;
    }

    final hubUrl = '${Environments.apiUrl}$_hubPath';
    debugPrint('[SignalR] connecting to $hubUrl');

    final connection = HubConnectionBuilder()
        .withUrl(
          hubUrl,
          options: HttpConnectionOptions(
            accessTokenFactory: _ensureAccessToken,
          ),
        )
        .withAutomaticReconnect(retryDelays: [0, 2000, 5000, 10000, 30000])
        .build();

    connection.on(_membershipChangedEvent, _onMembershipChanged);
    connection.onclose(({Exception? error}) {
      debugPrint('[SignalR] closed: $error');
    });
    connection.onreconnected(({String? connectionId}) {
      debugPrint('[SignalR] reconnected: $connectionId');
    });

    _connection = connection;
    await connection.start();
    debugPrint('[SignalR] connected');
  }

  Future<void> disconnect() async {
    final connection = _connection;
    _connection = null;
    if (connection == null) return;

    try {
      connection.off(_membershipChangedEvent);
      await connection.stop();
    } catch (_) {}
  }

  Future<String> _ensureAccessToken() async {
    final current = await _tokenStorage.getAccessToken();
    if (current != null &&
        current.isNotEmpty &&
        !JwtDecoder.isExpired(current)) {
      return current;
    }

    final refreshToken = await _tokenStorage.getRefreshToken();
    if (refreshToken == null || refreshToken.isEmpty) {
      return current ?? '';
    }

    try {
      final response = await _refreshDio.post(
        '/auth/refresh',
        data: {'refreshToken': refreshToken},
      );
      final newAccess = response.data['accessToken'] as String;
      final newRefresh = response.data['refreshToken'] as String;
      await _tokenStorage.saveUserTokens(newAccess, newRefresh);
      return newAccess;
    } catch (e) {
      debugPrint('[SignalR] token refresh failed: $e');
      return current ?? '';
    }
  }

  void _onMembershipChanged(List<Object?>? args) {
    debugPrint('[SignalR] GroupMembershipChanged raw=$args');
    if (args == null || args.isEmpty) return;

    final event = GroupMembershipChanged.tryParse(args.first);
    if (event == null) {
      debugPrint('[SignalR] failed to parse membership event');
      return;
    }

    for (final handler in List<GroupMembershipHandler>.from(_handlers)) {
      handler(event);
    }
  }
}
